### DISCUSSION TITLE
check if Current OS support argument

### DISCUSSION THREAD

**Discussion by wnunezc** [cefsharp/CefSharp#4173] [General]
Any Way to check if Current OS support argument like:

```c#
CefSettings settings = new CefSettings();

settings.CefCommandLineArgs.Add("in-process-gpu");
settings.CefCommandLineArgs.Add("use-angle", "gl");
```

in mi dev Laptop with windows 11 it work well when i test the release executable in other PC with windows 10, it is automaticaly Closed.

**Comment by amaitland** [comment]
> Any Way to check if Current OS support argument like:

Not to my knowledge. You would have to manually test with all os/hardware combinations you wish to support.

Hard coding specific `GPU` settings is not recommended, it's best to make sure your app has an [app.manifest](https://github.com/cefsharp/CefSharp/blob/cefsharp/103/NuGet/Readme.txt#L13) and let `Chromium` decide what best suites your hardware.

Also it's my understanding that `in-process-gpu` is only for testing purposes on `Windows`. Use as your own risk.

If you are trying to specify `GPU command line args` for the purpose of taking a screenshot then that's entirely unnecessary now. Use https://cefsharp.github.io/api/102.0.x/html/M_CefSharp_WinForms_ChromiumWebBrowser_CaptureScreenshotAsync.htm

**Reply by wnunezc** [reply]
I am using it to enable the best engine option for webgl support and best FPS performance...
in a friend's windows 10 installation the program closes automatically... when you enable these two options...

**Reply by wnunezc** [reply]
I'm trying to create an application that allows me to open FlyFF universe which is a game based on webgl but I want it to run in the best way and with the best performance... I've had to do several tricks just for it to take advantage of the card dedicated video instead of using the integrated one... now I'm trying to improve the FPS with the engine that it implements for WebGL rendering

**Reply by amaitland** [reply]
Did you add an `app.manifest`? You'll need one for `GPU` to work correctly.

Using the command line args you suggested is not recommended.

> I've had to do several tricks just for it to take advantage of the card dedicated video instead of using the integrated one

You should either SelfHost the browser sub process or specify a priority to `CefSharp.BrowserSubprocess.exe`. Don't use in process GPU.

> now I'm trying to improve the FPS with the engine that it implements for WebGL rendering

No such changes should be necessary if you have an app.manifest.

**Reply by amaitland** [reply]
Test with https://github.com/cefsharp/CefSharp.MinimalExample 

Performance should be nearly identical to `Chromium` at the same version.

**Reply by wnunezc** [reply]
![image](https://user-images.githubusercontent.com/1207306/179445368-b795e099-669c-41ca-8b6e-91eb61d6f7e1.png)

**Reply by wnunezc** [reply]
i am using C# and winform

**Reply by amaitland** [reply]
`GPU` process won't behave properly with the example you've provided.

**Reply by wnunezc** [reply]
the minimal example not show how to implement Dedicated GPU Acceleration or best performance config for WebGL acceleration...

**Reply by wnunezc** [reply]
thanks anyway for the whispers, but I don't think there's anything detailed on how to do this... I'll follow the trial and error to see if I find something useful...

**Reply by wnunezc** [reply]
and i think that cant implemente app.manifest to handle it becouse i am using .net 4.5.2 and that you say is working in .net 4.7 or higger

**Reply by amaitland** [reply]
Have you compared the performance of the `MinimalExample`? You `app.manifest` must include the relevant compatibility entries, the `MinimalExample` is provided as a reference.

As I've already said you need to tell your `GPU Software` that `CefSharp.BrowserSubProcess.exe` should use your dedicated `GPU`. 

> best performance config for WebGL acceleration

In most cases no additional configuration is required. Best performance configuration would also depend on the hardware/chromium version used. `Chromium` generally does an excellent job of working this out for itself.

chrome://gpu can be loaded in` CefSharp` and `Chrome` to compare settings.

If you wish to hard code settings then you'll have to deal with any problems that arise from that.

**Reply by amaitland** [reply]
> and i think that cant implemente app.manifest to handle it becouse i am using .net 4.5.2 and that you say is working in .net 4.7 or higger

Not true.

**Reply by wnunezc** [reply]
ok i am crazy...
![image](https://user-images.githubusercontent.com/1207306/179446620-5b954e44-99c1-4b62-8a3d-29237f7540ef.png)

**Reply by amaitland** [reply]
You can still configure `DPI Awareness` via the `app.manifest` in older versions of `.Net. Newer versions do have better support, doesn't mean that older versions have none.

https://github.com/cefsharp/CefSharp/wiki/General-Usage#appmanifest

**Reply by amaitland** [reply]
Use https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/app.manifest and change assemblyIdentity.

**Reply by wnunezc** [reply]
with my manifest:
![image](https://user-images.githubusercontent.com/1207306/179448903-87b2f8ac-e09b-4b2f-a682-cfe3aa3fc4fb.png)

With your manifest:
![image](https://user-images.githubusercontent.com/1207306/179448753-4e1511dc-7a03-43af-8668-5073c086c59c.png)

adding the manifest of your minimal example doesn't help me at all...even the frame rate drops abruptly...

**Comment by wnunezc** [comment]
@amaitland well I have tried to do my best and the main idea is to show this to the User:

![image](https://user-images.githubusercontent.com/1207306/179771429-60ea0ad4-a253-4f06-bc6e-f18363fed562.png)

As you can see I am offering the user the hardware acceleration that CefSharp offers and CEF...

```c#
CefSettings settings = new CefSettings();

settings.CefCommandLineArgs.Add("in-process-gpu");

switch (index)
{
	case 0:
		settings.CefCommandLineArgs.Add("use-angle", "gl");
	break;
	case 1:
		settings.CefCommandLineArgs.Add("use-angle", "gles");
	break;
	case 2:
		settings.CefCommandLineArgs.Add("use-angle", "d3d9");
	break;
	case 3:
		settings.CefCommandLineArgs.Add("use-angle", "d3d11");
	break;
	case 4:
		settings.CefCommandLineArgs.Add("enable-features", "Vulkan");
		settings.CefCommandLineArgs.Add("use-vulkan", "1");
	break;
}
```

the problem is that I need to test if the selected version works correctly at the moment the user selects one of them. and prevent the program from breaking and not being able to reopen...

**Reply by wnunezc** [reply]
the main reason that an option of this does not work; it is because the OS does not support it or CEF or CefSharp does not find locally the support to provide the selected hardware acceleration; I cannot create an executable for each version and have the end user download N amounts of executables that vary because an option cannot be tested or validated.
